### DISCUSSION TITLE
Room.Send 함수 잘 안돼요

### DISCUSSION THREAD

**Discussion by Deytooh** [naverz/zepeto-studio-kor#458] [Issue Report (Legacy)]
안녕하세요 ~ 

`interface MyInterface{
    myVar: string;
    myVarSecond: string;
}
`

이런 interface 있는데 -> 

` let message=
            {
                myVar: "wow",
                myVarSecond: "omg"
            } as MyInterface;
        this.room.Send("onInstantiateNetwork", message);`

이거 하고 Send 함수 하고 ->

서버에 이벤트 있어요 ->

`  this.onMessage<MyInterface>("onInstantiateNetwork", (client, message) => {
            console.log(message.myVar);
        });`

이거는, 잘 돼요 👍 
하지만...

`interface MyInterface{
    myVar: string;
   myPosX;number;
myPosY;number;
myPosZ;number;
myRotX;number;
myRotY;number;
myRotZ;number;
myRotX;number;
myScaleY;number;
myScaleZ;number;
    myVarSecond: string;
}
`

이거 하면, 안돼요. 서버 메세지 안와요 ~

**Comment by Zepeto-TechEvangelist** [comment]
안녕하세요, 말씀하신 내용이 인터페이스 내의 속성을 많이 선언하면 안된다는 뜻일까요?
관련해서 schema, index, clientstarter쪽 코드를 상세히 보여주시면 도움이 될 것 같습니다.

**Reply by Deytooh** [reply]
네 맞아요 ! Interface에서 속성 많으면 서버 메세지 안와요 ! 코드 똑같은데 속성 많지않으면 메세지 잘 와요 ! 

다름 방법으로 하고 interface 코드 다 삭제 해서 코드 없어요 ㅠㅠ 다음에 다시 할때 공유하겠습니다 !

**Reply by Zepeto-TechEvangelist** [reply]
넵~ 확인해보고 말씀드릴게요!

**Reply by Zepeto-TechEvangelist** [reply]
안녕하세요, 저희가 재현했을때는 인터페이스 속성이 많아도 잘 출력되었습니다. (23개까지 테스트 했습니다)
혹시 선언하실때 코드에서 
myPosY;number; myPosZ;number; 이렇게 세미콜론을 중간에 써서 오류가 났던게 아닌지 추측됩니다.

**Reply by Deytooh** [reply]
오오 진짜요 ?? 다시 해봐야겠네요 ! 혹시 잘되는 코드 보내줄 수 있을까요 ? 
아아 네 myPos;number 이거 그냥 github에서 메세지 할려고해서 빨리 다 바꿨지만 원래 이거였어요 -> 
` interface NetworkObjectMessage {
     prefabHash: string;
     pos_x: number;
     pos_y: number;
     pos_z: number;
     rot_x: number;
     rot_y: number;
     rot_z: number;
     scale_x: number;
     scale_y: number;
     scale_z: number;
     parentName: string;
 }`
 
 이걸로 안됐어요 !

**Reply by Zepeto-TechEvangelist** [reply]
![image](https://user-images.githubusercontent.com/100118650/159417440-5486f3c8-d354-4231-9f02-bcf9bad49163.png)

이렇게 테스트했었습니다.

**Reply by Deytooh** [reply]
오 저도 똑같았어요 !! 그럼 다시 테스트해보고 말씀드릴게요 !!

**Comment by shlifedev** [comment]
안녕하세요.  interface의 내용들을 `RoomData` 에 담아 전송해보시겠어요?

메세지 안에 객체가 포함되어 있거나 (Vector3 같은) 그 외의 케이스에서 제대로
메세지 내용이 전달 안되는 경우가 있는 것 같습니다.

이런 경우 RoomData로 전송시에는 문제 없이 보내지는 것 같습니다.

서버와 실제로 메세지를 주고받을때 json string이 아니라 이진데이터로 변환하여 보내게 되는데
내부에서 동작하는 방식의 문제로 보여집니다.

RoomData에 포함시키는게 좀 불편하시겠지만 한번 시도해보시면 좋을 것 같아요.
[멀티플레이](https://github.com/naverz/zepeto-multiplay-example) 예제에 있는 샘플코드를 남겨드립니다.

```ts
private SendTransform(transform: UnityEngine.Transform) {
        const data = new RoomData();
 
        const pos = new RoomData();
        pos.Add("x", transform.localPosition.x);
        pos.Add("y", transform.localPosition.y);
 
        pos.Add("z", transform.localPosition.z);
        data.Add("position", pos.GetObject());
 
        const rot = new RoomData();
        rot.Add("x", transform.localEulerAngles.x);
        rot.Add("y", transform.localEulerAngles.y);
        rot.Add("z", transform.localEulerAngles.z);
        data.Add("rotation", rot.GetObject());
        this.room.Send("onChangedTransform", data.GetObject());
    }
```
